Mesh constraint matrix support in FileMeshGenerator

roystgnr opened the pull request:
## Reason
This lets us combine an arbitrary extraction operator (expressed as a sparse matrix file) with a mesh of assembly elements, in a generalization of what we do with IsoGeometric Analysis meshes now that lets us support more complicated mesh types.

## Design
If a constraint matrix file is requested, we load it and import its constraints along with the mesh file it applies to; doing this in FileMeshGenerator rather than a separate generator ensures we can control renumbering settings and make sure the ordering from the files stays matched.

## Impact
No impact for users who don't enable the option; more mesh types available for users who do.

Oh, except I also slipped in a bug-fix/optimization that I noticed while setting up a regression test using SideSetsFromNormalsGenerator

Refs #18768

roystgnr committed:
Add MatrixFileName MOOSE type

FileName would be adequate here, but the docs for MeshFileName suggest
the possibility of being more intelligent about file extensions in the
future and if we want to do that there we'll probably want to here.

roystgnr committed:
Prerequest from libMesh::FE

This fixes an assertion failure with --disable-deprecated libMesh
builds, and improves efficiency with other builds.

roystgnr committed:
Add FileMeshGenerator constraint_matrix option

Refs #18768

This lets us run with IGA meshes using more flexible extraction
operators than the body-fit meshes we've been restricted to previously.

roystgnr committed:
Add iga_constraint_matrix regression test

This exercises the separate-constraint-matrix functionality.

roystgnr committed:
Use read_matlab() for now

libMesh::SparseMatrix::read() didn't make it to a submodule update yet.

moosebuild commented:
Job [Documentation](https://civet.inl.gov/job/2169789/) on 94c0d8a wanted to post the following:

View the site [here](https://mooseframework.inl.gov/docs/PRs/27345/site)

This comment will be updated on new commits.

roystgnr committed:
Prerequest properly in *all* SideSets generators

Okay, it's probably time to just disallow the old inefficient behavior
from the libMesh side, regardless of whether it's a --disable-deprecated
build or not.

roystgnr committed:
Respect (and document) allow_renumbering overrides

roystgnr commented:
Those remaining failures look real - I'm seeing different exodiff results on different processor counts, larger than roundoff error could explain - but I'm not able to reproduce them in the read/constrain/solve unit tests in https://github.com/libMesh/libmesh/pull/3831, so I'd wondering if I'm doing something wrong in our ExodusII output, but in paraview the diffs look too smooth to just be some sort of straightforward misindexing issue.

Looks like I can reproduce the problem in 1D, which should be simple enough to debug by hand.

roystgnr committed:
cache_elem_data() if needed to stay prepared

This is much faster than setting unprepared and making MOOSE do the
whole prepare_for_use() later.  My mistake not to do it at the libMesh
level, but we'll fix that in a later submodule update.

roystgnr committed:
Fix flood() with distributed meshes

roystgnr committed:
Use fixed_normal=true in test

We only need this to define the sidesets we want, and in theory this is
distributed-mesh-compatible.

GiudGiud reviewed (commented):
I would create a parameter group for the IGA parameters now since there are a few with
```
params.addParamNameToGroup(params, "group")
```

A bit of doco in FileMeshGenerator.md would help too?

roystgnr committed:
Give IGA parameters their own group

Thanks to Guillaume for the suggestion

moosebuild commented:
Job [Coverage](https://civet.inl.gov/job/2359215/) on c6d2b4e wanted to post the following:


## Framework coverage

<table>
  <thead>
    <tr>
      <th></th>
      <th>89ffe9</th>
      <th colspan=4>#27345 c6d2b4</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>85.07%</td>
      <td>85.07%</td>
      <td>+0.00%</td>
      <td>97.06%</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>104700</td>
      <td>104726</td>
      <td>+26</td>
      <td>33</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>18374</td>
      <td>18376</td>
      <td>+2</td>
      <td>1</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/27345/coverage_diff/framework)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/27345/coverage/framework)


## Modules coverage

Coverage did not change

## Full coverage reports

<details>
<summary>Reports</summary>

- [`framework`](https://mooseframework.inl.gov/docs/PRs/27345/coverage/framework)
- [`chemical_reactions`](https://mooseframework.inl.gov/docs/PRs/27345/coverage/chemical_reactions)
- [`combined`](https://mooseframework.inl.gov/docs/PRs/27345/coverage/combined)
- [`contact`](https://mooseframework.inl.gov/docs/PRs/27345/coverage/contact)
- [`electromagnetics`](https://mooseframework.inl.gov/docs/PRs/27345/coverage/electromagnetics)
- [`external_petsc_solver`](https://mooseframework.inl.gov/docs/PRs/27345/coverage/external_petsc_solver)
- [`fluid_properties`](https://mooseframework.inl.gov/docs/PRs/27345/coverage/fluid_properties)
- [`fsi`](https://mooseframework.inl.gov/docs/PRs/27345/coverage/fsi)
- [`functional_expansion_tools`](https://mooseframework.inl.gov/docs/PRs/27345/coverage/functional_expansion_tools)
- [`geochemistry`](https://mooseframework.inl.gov/docs/PRs/27345/coverage/geochemistry)
- [`heat_transfer`](https://mooseframework.inl.gov/docs/PRs/27345/coverage/heat_transfer)
- [`level_set`](https://mooseframework.inl.gov/docs/PRs/27345/coverage/level_set)
- [`misc`](https://mooseframework.inl.gov/docs/PRs/27345/coverage/misc)
- [`navier_stokes`](https://mooseframework.inl.gov/docs/PRs/27345/coverage/navier_stokes)
- [`optimization`](https://mooseframework.inl.gov/docs/PRs/27345/coverage/optimization)
- [`peridynamics`](https://mooseframework.inl.gov/docs/PRs/27345/coverage/peridynamics)
- [`phase_field`](https://mooseframework.inl.gov/docs/PRs/27345/coverage/phase_field)
- [`porous_flow`](https://mooseframework.inl.gov/docs/PRs/27345/coverage/porous_flow)
- [`ray_tracing`](https://mooseframework.inl.gov/docs/PRs/27345/coverage/ray_tracing)
- [`rdg`](https://mooseframework.inl.gov/docs/PRs/27345/coverage/rdg)
- [`reactor`](https://mooseframework.inl.gov/docs/PRs/27345/coverage/reactor)
- [`richards`](https://mooseframework.inl.gov/docs/PRs/27345/coverage/richards)
- [`scalar_transport`](https://mooseframework.inl.gov/docs/PRs/27345/coverage/scalar_transport)
- [`solid_mechanics`](https://mooseframework.inl.gov/docs/PRs/27345/coverage/solid_mechanics)
- [`solid_properties`](https://mooseframework.inl.gov/docs/PRs/27345/coverage/solid_properties)
- [`stochastic_tools`](https://mooseframework.inl.gov/docs/PRs/27345/coverage/stochastic_tools)
- [`thermal_hydraulics`](https://mooseframework.inl.gov/docs/PRs/27345/coverage/thermal_hydraulics)
- [`xfem`](https://mooseframework.inl.gov/docs/PRs/27345/coverage/xfem)

</details>



This comment will be updated on new commits.

lindsayad commented on framework/src/meshgenerators/FileMeshGenerator.C:
```diff
@@ -70,6 +77,14 @@ FileMeshGenerator::generate()
+
+  // If we have a constraint matrix, we need its numbering to match
+  // the numbering in the mesh file
+  if (_matrix_file_name != "")
```
```suggestion
  if (!_matrix_file_name.empty())
```

lindsayad commented on framework/src/meshgenerators/FileMeshGenerator.C:
```diff
@@ -129,6 +145,25 @@ FileMeshGenerator::generate()
     _app.possiblyLoadRestartableMetaData(MooseApp::MESH_META_DATA, (std::string)file_name);
   }
 
+  if (_matrix_file_name != "")
```
```suggestion
  if (!_matrix_file_name.empty())
```

lindsayad commented on test/tests/meshgenerators/file_mesh_generator/tests:
```diff
@@ -157,6 +157,38 @@
     issues = '#18768'
   []
 
+  [./iga_constraint_matrix]
```
```suggestion
  [iga_constraint_matrix]
```

lindsayad commented on test/tests/meshgenerators/file_mesh_generator/tests:
```diff
@@ -157,6 +157,38 @@
+    issues = '#18768'
+  []
+
+  [./iga_constraint_matrix_1d]
```
```suggestion
  [iga_constraint_matrix_1d]
```

roystgnr committed:
Apply suggestions from code review

Co-authored-by: Alex Lindsay <alexlindsay239@gmail.com>
